sklearn/tree/tree.py

```diff
@@ -118,9 +118,16 @@ def fit(self, X, y, sample_weight=None, check_input=True,
             X_idx_sorted=None):
 
         random_state = check_random_state(self.random_state)
+        accept_sparse = None
+        sparse_formats = ['bsr', 'csr', 'csc', 'coo', 'dia', 'dok', 'lil']
+        is_classification = isinstance(self, ClassifierMixin)
+        if is_classification:
+            accept_sparse = sparse_formats
+
         if check_input:
             X = check_array(X, dtype=DTYPE, accept_sparse="csc")
-            y = check_array(y, ensure_2d=False, dtype=None)
+            y = check_array(y, ensure_2d=False, dtype=None,
+                            accept_sparse=accept_sparse)
```

Comment:
I added this to ensure a consistent error message for the classification case. Since we are to change the message, I'll revert this.